episodes/5-transfer-learning.Rmd

```diff
@@ -0,0 +1,235 @@
+```python
+base_model.trainable = False
+```
+
+Let's define our 'head' network:
+```python
+out = base_model.output
+out = keras.layers.Flatten()(out)
+out = keras.layers.BatchNormalization()(out)
+out = keras.layers.Dense(128, activation='relu')(out)
+out = keras.layers.Dropout(0.5)(out)
+out = keras.layers.BatchNormalization()(out)
+out = keras.layers.Dense(64, activation='relu')(out)
+out = keras.layers.Dropout(0.5)(out)
+out = keras.layers.Dense(10)(out)
+```
+
+Finally we define our model:
+```python
+model = keras.models.Model(inputs=inputs, outputs=out)
+```
+::: challenge
+## Inspect the DenseNet121 network
+Have a look at the network architecture with `model.summary()`.
+It is indeed a deep network, so expect a long summary!
+
+### 1.Trainable parameters
+How many parameters are there? How many of them are trainable? 
+
+Why is this and how does it effect the time it takes to train the model?
+
+### 2. Head and base
+Can you see in the model summary which part is the base network and which part is the head network?
+
+### 3. Max pooling
+Which layer is added because we provided `pooling='max'` as argument for `DenseNet121()`?
+
+:::: solution
+## Solutions
+### 1. Trainable parameters
+Total number of parameters: 7182218, out of which only 142410 are trainable.
+
+The 142410 trainable parameters are the weights of the head network. All other parameters are 'frozen' because we set `base_model.trainable=False`. Because only a small proportion of the parameters have to be updated at each training step, this will greatly speed up training time.
+
+### 2. Head and base
+The head network starts at the `flatten` layer, 8 layers before the final layer.
+
+### 3. Max pooling
+The `max_pool` layer right before the `flatten` layer is added because we provided `pooling='max'`.
+::::
+:::
+
+
+
+::: challenge
+## Training and evaluating the pre-trained model
+
+### 1. Compile the model
+Compile the model:
+- Use the `adam` optimizer 
+- Use the `SparseCategoricalCrossentropy` loss with `from_logits=True`. 
+- Use 'accuracy' as a metric.
+
+### 2. Train the model
+Train the model on the training dataset:
+- Use a batch size of 32
+- Train for 30 epochs, but use an earlystopper with a patience of 5
+- Pass the validation dataset as validation data so we can monitor performance on the validation data during training
+- Store the result of training in a variable called `history`
+- Training can take a while, it is a much larger model than what we have seen so far.
+
+### 3. Inspect the results
+Plot the training history and evaluate the trained model. What do you think of the results?
+
+
+:::: solution
+## Solution
+
+### 1. Compile the model
+```python
+model.compile(optimizer='adam',
+              loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+              metrics=['accuracy'])
+```
+
+### 2. Train the model
+Define the early stopper:
+```python
+keras.callbacks.EarlyStopping(monitor='val_accuracy',
```

Comment:
I understand you need to store the `EarlyStopping` object in a variable to use it in the next step.

```suggestion
early_stopper = keras.callbacks.EarlyStopping(monitor='val_accuracy',
```

Reply:
Whoops!